### ISSUE TITLE
DOMException thrown because "The user denied permission to use Service Worker." in sdk context and incognito mode of google-chrome

### ISSUE THREAD

**Issue by trepmag** [unep-grid/mapx#580]
This was observed while using google-chrome version 84 when in incognito mode; example: https://observablehq.com/@trepmag/mapx-sdk-introduction?collection=@trepmag/mapx-sdk

```
?project=MX-YBJ-YYF-08R-UUR-QW6&language=en&lockProject=true:1
Uncaught (in promise) DOMException:
  Failed to register a ServiceWorker for scope ('https://app.mapx.org/') with script ('https://app.mapx.org/service-worker.js'):
    The user denied permission to use Service Worker.
```

Couldn't reproduce this while using chromium version 84 of firefox...

I'm reporting this while this might be an issue with google-chrome incognito mode and service-worker when embedded...

**Comment by fxi** [comment]
*or firefox ?

chromium linux only ?

We do this test to check if everything is compatible:

```js
const isCompatible =
  'serviceWorker' in navigator &&
  'storage' in navigator &&
  'estimate' in navigator.storage &&
  'caches' in window;
``

There is anything else to check ? I have no idea what to test for ...

thanks

**Comment by trepmag** [comment]
- It fails with google-chrome windows + linux (version 84)
- It works with chromium and firefox (linux tested only)

**Comment by trepmag** [comment]
@fxi , does it works on you mac google-chrome browser?

**Comment by fxi** [comment]
In incognito mode, nope. 

Could you investigate how could we check for this ?  ` 'serviceWorker' in navigator ` is not enough..

Thanks.

**Comment by trepmag** [comment]
Yes will do...

**Comment by trepmag** [comment]
With local dev environment the issue appens in `getBrowserData()` function while accessing `window.sessionStorage`; this thow in mx_helper_misc.js (https://github.com/unep-grid/map-x-mgl/blob/master/app/src/js/mx_helper_misc.js#L1997):
```
Uncaught DOMException: Failed to read the 'sessionStorage' property from 'Window': Access is denied for this document.
```

Now, altering `getBrowserData()` as follow trigger another exception later in the execution flow ;-(:
```js
export function getBrowserData() {

  let hasSessionStorage = false;
  try {
    hasSessionStorage = !!window.sessionStorage;
  } catch(e) {}

  return {
    language: navigator.language.substr(0, 2),
    cookies: mx.helpers.readCookie(),
    userAgent: navigator.userAgent,
    timeZone: new Date().toString().replace(/.*[(](.*)[)].*/, '$1'),
    hasLocalStorage: hasSessionStorage,
    hasSessionStorage: hasSessionStorage,
    hasGeolocation: !!navigator.geolocation
  };
}
```

Furthermore, I test in a simple example outside mapx to access `window.sessionStorage` in incognito mode with google-chrome in a iframe and it return `Storage {length: 0}` as opposed to the exception seen above in mapx context which is build upon 'shiny:connected'... Well, I couldn't figure a solution in a reasonable time...

**Comment by fxi** [comment]
Is this another issue or it's related to the permission to use service worker ? I'm not sure to understand.

**Comment by trepmag** [comment]
It might be related issues.

In prod it throws:
```
Uncaught (in promise) DOMException:
  Failed to register a ServiceWorker for scope ('https://app.mapx.org/') with script ('https://app.mapx.org/service-worker.js'):
    The user denied permission to use Service Worker.
```

In local it throws:
```
Uncaught DOMException:
  Failed to read the 'sessionStorage' property from 'Window': 
    Access is denied for this document.
```

**Comment by fxi** [comment]
Blocked IP request seems to be the more reliable way to know if the browser is in some sort of security/private mode. 

So, this step could be done before trying to register service workers, but it requires to wait on the result before going further, which will block the app loading. 

Here is the draft of a function to indirectly test if the browser is in security/private mode.

```js
async function testPrivate() {
  let url = 'https://api.mapx.org/get/ip';
  try {
    let r = await fetch(url);
    let data = await r.json();
    console.log('net ip ok: non private :)');
    return true;
  } catch (e) {
    console.log('net error, private  ?', e);
    return false;
  }
}
```

@trepmag What's your thoughts about this ?

**Comment by thomaspiller** [comment]
It seems that this is no longer a problem in the latest version of the browers. @trepmag could you please check this and close the issue if so? Thanks

**Comment by thomaspiller** [comment]
According to my tests, this problem is no longer reproducible. I am closing this issue.
